[clang][test] Add test for comma operator rejection in preprocessor conditionals

swote-git opened the pull request:
Add test coverage to ensure comma operators remain properly rejected in `#if` directives.

Per CWG 1436, comma is not among the permitted operators in preprocessor conditional expressions. This test prevents regressions and clarifies the intended behavior.

Fixes #132822

swote-git committed:
[clang][test] Add tests for comma operator rejection in preprocessor expressions

Add test coverage for comma operator usage in #if preprocessor directives
to ensure it continues to be properly rejected across all C++ standard versions.
Per CWG 1436, comma operators are not among the permitted operators in
preprocessor conditional expressions.

Addresses #132822

swote-git commented:
Currently, this only about adding test code.

While adding these tests, I noticed two potential improvements to the current error message `"comma operator in operand of #if"`:

1. The same message appears for `#elif`, `#ifdef`, etc.
2. Per CWG 1436, comma operators are explicitly **not permitted** in preprocessor conditionals. The current message describes **where** the comma is, but doesn't clearly convey **why** it's prohibited.

A clearer message should be more clear.

Sirraide reviewed (commented):
LGTM, though I wonder if it should be part of the DR tests instead.

CC @Endilll

DavidSpickett commented on clang/test/Preprocessor/cxx_oper_comma.cpp:
```diff
@@ -0,0 +1,26 @@
+// Test 1: Top-level comma
+// expected-error@+1 {{expected end of line in preprocessor expression}}
+#if 1, 2
+#endif
```
I presume our preprocessor treats `#elif` exactly the same, as `<keyword> <expression>` however I would check at least one of these with `#elif` as well. Just to be sure that's how it works.

swote-git replied:
Missing test case for #elif needs to be corrected. 

Thank you for your thoughtful review!

DavidSpickett commented:
There are some tests in `clang/test/Preprocessor/expr_invalid_tok.c` right now.

swote-git commented:
@DavidSpickett Thanks for your help.

I can see similar test in `expr_invalid_tok.c`

```c
// PR2284 - The constant-expr production does not including comma.
// CHECK: [[@LINE+1]]:14: error: expected end of line in preprocessor expression
#if 1 ? 2 : 0, 1
#endif
```

What I wanted to write a little more comprehensively was because of the contents of `expr_comma.c`.

```c
// Comma is not allowed in C89
// RUN: not %clang_cc1 -E %s -std=c89 -pedantic-errors

// Comma is allowed if unevaluated in C99
// RUN: %clang_cc1 -E %s -std=c99 -pedantic-errors 

// PR2279

#if 0? 1,2:3
#endif
```

swote-git committed:
Add #elif test and CWG 3017 context

- Add test coverage for comma operator in #elif directive
- Reference CWG 3017 in comments for additional context
- Addresses reviewer feedback

swote-git committed:
Add test for #if , (leading comma syntax error)

Endilll commented on clang/test/Preprocessor/cxx_oper_comma.cpp:
```diff
@@ -0,0 +1,38 @@
+// RUN: %clang_cc1 -E -pedantic-errors %s -verify -std=c++23
+
+// Test 1: Top-level comma
+// expected-error@+1 {{expected end of line in preprocessor expression}}
```
Can you open an issue to improve diagnostics here? It would be more actionable to point at the comma than to say that EOL was expected

Endilll commented on clang/test/Preprocessor/cxx_oper_comma.cpp:
```diff
@@ -0,0 +1,38 @@
+// RUN: %clang_cc1 -E -pedantic-errors %s -verify -std=c++14
+// RUN: %clang_cc1 -E -pedantic-errors %s -verify -std=c++17
+// RUN: %clang_cc1 -E -pedantic-errors %s -verify -std=c++20
+// RUN: %clang_cc1 -E -pedantic-errors %s -verify -std=c++23
```
`-std=c++2c` is missing

Endilll commented on clang/test/Preprocessor/cxx_oper_comma.cpp:
```diff
@@ -0,0 +1,38 @@
+// Test 6: Leading comma (syntax error)
+// expected-error@+1 {{invalid token at start of a preprocessor expression}}
+#if ,
+#endif
```
Missing newline at the end

Endilll commented on clang/test/Preprocessor/cxx_oper_comma.cpp:
```diff
@@ -0,0 +1,38 @@
+#if 0
+#elif (1, 2) // expected-error {{comma operator in operand of #if}}
+#endif
+
```
`limit` parameter of `#embed` directive is also evaluated as constant-expression https://eel.is/c++draft/cpp#embed.param.limit-2. I think it's worth adding a test case for this, too.

swote-git replied:
I added test for #embed case.

 For now, error comes up `expected ')'` , i'm not sure this is right diagnostics.

Endilll reviewed (commented):
Thank you for your work

swote-git committed:
Add test for #embed limit as a constant expression

swote-git committed:
adding new line

swote-git committed:
Address reviewer feedback for #embed test

Remove defined(__has_embed) check and use since-cxx23 verify prefix to align with C++ DR test conventions.

Endilll reviewed (commented):
I think this is good to go after you add `#if ,` test suggested earlier

swote-git commented:
Regarding the #if , test case, I believe I've already added it in the previous version of the patch(test 6). You can find it right here on line[36-39](https://github.com/llvm/llvm-project/pull/155570/files#diff-0339f6a182c253b6ed4d5239b390cc8653cf07187e50e5a3baa9e44b815d3258R37).

Thanks for your support!
